tensorboard/defs/web.bzl

```diff
@@ -161,21 +165,30 @@ def _tf_web_library(ctx):
             files=ts_files,
         ).to_json())
     er_config = _new_file(ctx, "-tsc-execroot.json")
-    ctx.file_action(output=er_config, content=execroot.to_json())
+    ctx.actions.write(output=er_config, content=execroot.to_json())
     ts_inputs = depset(
         [ts_config, er_config],
         transitive=[
             ts_inputs,
+            # TODO(@wchargin): Because `_tsc` is passed as `tools`
+            # below, collecting its runfiles should in principle be
+            # unnecessary, but without this line the two deps of `_tsc`
+            # (`tsc.js` and `@org_nodejs//:bin/node`) are in fact not
+            # included, and so execrooter fails at runtime.
             collect_runfiles([ctx.attr._tsc]),
             ts_typings,
             ts_typings_execroots,
         ],
     )
-    ctx.action(
-        inputs=list(ts_inputs),
+    ctx.actions.run(
+        tools=ctx.files._tsc,
```

Comment:
It’s a [list of files provided by the `tsc` target][doc]. That target is
[a `sh_binary`][tsc], so its output file is just the `tsc` binary. The
target’s `data` deps include its proper transitive dependencies, which
[should be sufficient to include them in `tsc`’s runfiles][data], and
the point of the `tools` kwarg is that [runfiles are automatically made
available to the action][tools]. Hence my confusion.

[doc]: https://docs.bazel.build/versions/master/skylark/lib/ctx.html#files
[tsc]: https://github.com/tensorflow/tensorboard/blob/b90a9a0ab7d27c044e58418bb2f34d9ea4d3fb2a/third_party/js.bzl#L103-L110
[data]: https://docs.bazel.build/versions/master/be/common-definitions.html#common.data
[tools]: https://docs.bazel.build/versions/master/skylark/lib/actions.html#run.tools

Reply:
@c-parsons, can you shed any light here?

Reply:
... which part of this are you hoping I can shed some light on? :) (can you clarify your question?)

Reply:
Sorry, yes. I’m passing `_tsc` as a tool to this `run` action, and
yet its two dependencies are not being included in the action’s
runfiles, so I have to manually include them as inputs.

Repro: Check out TensorBoard master (51fb0795), and apply the following
patch:

```diff
diff --git a/tensorboard/defs/web.bzl b/tensorboard/defs/web.bzl
index ced08234..2500fff9 100644
--- a/tensorboard/defs/web.bzl
+++ b/tensorboard/defs/web.bzl
@@ -170,18 +170,12 @@ def _tf_web_library(ctx):
         [ts_config, er_config],
         transitive=[
             ts_inputs,
-            # TODO(@wchargin): Because `_tsc` is passed as `tools`
-            # below, collecting its runfiles should in principle be
-            # unnecessary, but without this line the two deps of `_tsc`
-            # (`tsc.js` and `@org_nodejs//:bin/node`) are in fact not
-            # included, and so execrooter fails at runtime.
-            collect_runfiles([ctx.attr._tsc]),
             ts_typings,
             ts_typings_execroots,
         ],
     )
     ctx.actions.run(
-        tools=ctx.files._tsc,
+        tools=ctx.files._tsc + ctx.files._tsc_deps,
         inputs=ts_inputs,
         outputs=ts_outputs,
         executable=ctx.executable._execrooter,
@@ -407,6 +401,14 @@ tf_web_library = rule(
             allow_files=True,
             executable=True,
             cfg="host"),
+        "_tsc_deps": attr.label_list(
+            default=[
+                Label("@com_microsoft_typescript//:tsc.js"),
+                Label("@org_nodejs//:bin/node"),
+            ],
+            allow_files=True,
+            cfg="host",
+        ),
         "_default_typings": attr.label(
             default=Label("//tensorboard:tf_web_library_default_typings"),
             allow_files=True),
```

Then run `bazel build //tensorboard`, which works. But then delete the
two added `Label` lines (so that no additional deps are being passed),
and re-run `bazel build //tensorboard`, which fails with

```
bazel-out/host/bin/external/com_microsoft_typescript/tsc: line 6: /HOMEDIR/.bazel-output-base/sandbox/linux-sandbox/774/execroot/org_tensorflow_tensorboard/external/org_nodejs/bin/node: No such file or directory
```

(a failure in our tool, yes, but because the required runfiles are not
being provided). Verify that the dependencies are actually declared:

```
$ bazel query '@com_microsoft_typescript//:tsc' --output=build | san
# /HOMEDIR/.bazel-output-base/external/com_microsoft_typescript/BUILD:17:1
sh_binary(
  name = "tsc",
  data = ["@com_microsoft_typescript//:tsc.js", "@org_nodejs//:org_nodejs"],
  srcs = ["@com_microsoft_typescript//:tsc.sh"],
)
$ bazel query 'somepath(@com_microsoft_typescript//:tsc, @org_nodejs//:bin/node)'
@com_microsoft_typescript//:tsc
@org_nodejs//:org_nodejs
@org_nodejs//:bin/node
```

My question is: I thought that runfile dependencies of `tools` were
supposed to be included in the runfiles tree, and I thought that `data`
dependencies of `sh_binary` rules counted as runfiles dependencies, so
why must we include these deps manually?

Reply:
Your understanding matches mine, here. In terms of your specific code, I would expect that you need to include the `files_to_run` of your dependency to get the full runfiles of the tool itself, instead of just the executable file. So instead of:
```
tools=ctx.files._tsc
```

I would expect:
```
tools=[ctx.attr._tsc[DefaultInfo].files_to_run]
```

I tried this, however, and it doesn't work. I'll need to do some further digging.

Reply:
Okay; I’ve opened an issue on the Bazel repository for tracking:
repo for tracking:

<https://github.com/bazelbuild/bazel/issues/8697>

Thank you for looking into this!